rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,396 @@
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_SRC, mask);
+    attr.set_scales(DNNL_ARG_DST, mask);
+    ```
+  - `attr::rnn_weights_qparams`
+    ``` cpp
+    primitive_attr attr;
+    attr.set_scales(DNNL_ARG_WEIGHTS, mask);
+    ```
+  - `attr::p_op::built-in parameter`
+    ``` cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+    p_ops.append_eltwise(algorithm::eltwise_relu, alpha, beta);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(1), /* mask */ 0);
+
+    ...
+    std::unordered_map<int, memory> args;
+    ...
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SCALE | DNNL_ARG_SRC_1, p_op_0_src_1_scale_mem});
+    args.insert({DNNL_ARG_ATTR_MULTIPLE_POST_OP(1) | DNNL_ARG_SCALE, p_op_1_scale_mem});
+    p.execute(s, args);
+    ```
+ - `built-in parameter` in Sum
+   ``` cpp
+   primitive_attr attr;
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 0, /* mask */ 0);
+   attr.set_scales(DNNL_ARG_MULTIPLE_SRC + 1, /* mask */ 0);
+
+   ...
+   std::unordered_map<int, memory> args;
+   ...
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 0, scale_0_mem});
+   args.insert({DNNL_ARG_MULTIPLE_SRC + 1, scale_1_mem});
+   p.execute(s, args);
+   ```
+
+
+#### Simplifying scales configuration
+
+Currently correspondence mask defines the correspondence between the tensor
+dimensions and the scales vector. The set i-th bit indicates that a dedicated
+scaling factor is used for each index along that dimension. The mask value of 0
+implies a common scaling factor for the whole tensor.
+
+The are a few problems:
+1. Setting a mask is complicated. For example, to set per channels scales, mask
+   should be set to `2` (`1 << 1`) for a tensor where channels is a second
+   dimension.
+2. The same scale configuration should be set differently for tensor with
+   different semantics. For example, per channel scales:
+     - rnn data: `1 << 2`, because in rnn dimension order is is `tnc`, so
+       channels is 2nd dimension.
+     - conv data: `1 << 1`, because in conv dimension order is `nchw`, so
+       channels is 1st dimension.
+
+##### Improvement 1
+
+* This option doesn't break API, but simplifies mask usage.
+* oneDNN provides set of pre-defined correspondence masks:
+  ```c
+  // include/oneapi/dnnl/dnnl_types.h
+
+  #define DNNL_MASK_COMMON 0
+  #define DNNL_MASK_PER_DIM_0 1 << 0
+  #define DNNL_MASK_PER_DIM_1 1 << 1
+  #define DNNL_MASK_PER_DIM_2 1 << 2
+  ```
+  ```cpp
+  primitive_attr attr;
+  attr.set_scales(DNNL_ARG_DST, DNNL_MASK_COMMON);
+  ```
+
+##### Improvement 2
+
+* This option extends mask values to express dimensions based on dimension order
+  semantic.
+* oneDNN provides set of pre-defined correspondence masks:
+  ```c
+  // include/oneapi/dnnl/dnnl_types.h
+
+  #define DNNL_MASK_PER_CHANNEL_NCHW 1 << 12
```

Comment:
why is it not ?
```cpp
 #define DNNL_MASK_PER_CHANNEL_NCHW DNNL_MASK_PER_DIM_1
 #define DNNL_MASK_PER_CHANNEL_TNC DNNL_MASK_PER_DIM_2
 #define DNNL_MASK_PER_CHANNEL_OIHW DNNL_MASK_PER_DIM_0
```

Reply:
The main issue with such approach is that user still can pass DNNL_MASK_PER_CHANNEL_TNC to convolution and will not see the issue. Special values like `1 << 12` would help oneDNN to catch such issues at creation time.

This feature doesn't go into v3.0 due to low priority. This can be added later on demand since it doesn't break the API.